### PR TITLE
Adversarial ML Tools + documentation

### PR DESCRIPTION
**Pull request by lucasrussell01**
This PR contains scripts used for the adversarial training dataset production and evaluation that was used in v2p5, as well as some documentation. 
Added files
- "EventDrop.py" can be used to drop events from ROOT files to avoid large weights for very poorly balanced spectra.
- "EventMix.py" and "DataMixer.h" are used to mix the data and MC contributions together to generate the ROOT adversarial datasets.
- "AdversarialGenerate_tf.py" calls DataLoader with the new "GenerateAdversarial" option to assign pT weights and data/MC labels to taus when converting adversarial dataset to tensorflow format.
- "apply_adversarial_training.py" evaluates models on the adversarial testing dataset and saves the Djet scores for data/MC events (and y_adv for adversarial trained models) in the artifacts
- Jupyter notebook "evaluate_adversarial" to plot the adversarial predictions
- README_adversarial with some brief documentation of the steps

### CODE REVIEW

File: Training/configs/training_v1.yaml [hunk 1]
```diff
@@ -20,14 +20,16 @@ Setup:
...
-    input_type           : "ROOT" # supported "Adversarial", 'ROOT' or 'tf'
+    input_type           : "ROOT" # supported: 'ROOT', 'tf', 'Adversarial', 'AdversarialGenerate'
     tf_input_dir         : /eos/cms/store/group/phys_tau/TauML/prod_2018_v2/tf_datasets/emb_21k_99cut # path to saved tf dataset
     tf_dataset_x_order   : ["TauFlat", "inner_egamma", "inner_muon", "inner_hadron", "outer_egamma", "outer_muon", "outer_hadron"] # specify order of layers in tf dataset
     adversarial_dataset  : "/eos/cms/store/group/phys_tau/TauML/prod_2018_v2/adversarial_datasets/AdvTraining" # False if not use, ds name if use
     adv_parameter        : [1, 10] # k1, k2 for gradients in common layers
     n_adv_tau            : 100 # number of candidates per adversarial batch
     adv_learning_rate    : 0.01
     use_previous_opt     : False
+    adversarial_weights  : "/home/russell/AdversarialTauML/TauMLTools/adv_weights.root" # weights for adversarial dataset generation
```
Comment (kandrosov) [inline Training/configs/training_v1.yaml:31]: move weights to the central eos area and fix the path accordingly

File: Training/interface/DataLoader_main.h [hunk 2]
```diff
@@ -192,7 +205,7 @@ class DataLoader {
       if (!target_th2d) throw std::runtime_error("Target histogram could not be loaded");
 
       for( auto const& [tau_type, tau_name] : tau_types_names)
-      {
+      { if (tau_type != 8){ // No histogram for data samples
```
Comment (kandrosov) [inline Training/interface/DataLoader_main.h:208]: - fix indentations 
- use enum instead of 8 (also below)

File: Training/interface/DataLoader_main.h [hunk 3]
```diff
@@ -320,6 +343,13 @@ class DataLoader {
 
       }
 
+      const double GetAdversarialWeight(const ULong64_t dataset_id, const double pt) const
```
Comment (kandrosov) [inline Training/interface/DataLoader_main.h:346]:
```suggestion
      const double GetAdversarialWeight(const ULong64_t dataset_id, const double pt) const
```
```suggestion
      double GetAdversarialWeight(ULong64_t dataset_id, double pt) const
```

File: Training/interface/EventDrop.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
+#include "TFile.h"
```
Comment (kandrosov) [inline Training/interface/EventDrop.h:1]: a lot of code duplication between `EventDrop.h` and `DataMixer.h`. I suggest adding option to the `DataMixer` whatever to mix or to drop events and unify the code base

File: Training/python/EventMix.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
+import ROOT
```
Comment (kandrosov) [inline Training/python/EventMix.py:1]: `EventMix.py` -> `AdversarialEventMix.py`

File: Training/python/EventMix.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+n_tt_jets = 1
+n_DY_muons = 1
+
+print("Mix Characteristics:")
+
+
+dataDesc = ROOT.DatasetDesc()
+dataDesc.n_per_batch = n_data
+dataDesc.selection = "data"
+dataDesc.fileNames.push_back("/home/russell/skimmed_tuples/MuTau_prod2018/SingleMuon_skimmed.root")
```
Comment (kandrosov) [inline Training/python/EventMix.py:34]: move input root files to the cenral eos storage and fix paths accordingly (here and below)

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Training/interface/EventDrop.h (hunk 4 above)
Comment (lucasrussell01) [reply Training/interface/EventDrop.h]: I have added an option so that only `DataMixer.h` is necessary, please let me know if you don't like the way I've implemented it or have any suggestions for improvement

File: Training/interface/DataLoader_main.h (hunk 2 above)
Comment (lucasrussell01) [reply Training/interface/DataLoader_main.h]: I've now avoided using 8 and match to "data", by using tau_name for the first case and the unordered map find for the second, assuming this is ok

File: Training/interface/DataLoader_main.h [hunk 7]
```diff
@@ -204,23 +204,23 @@ class DataLoader {
...
-            ("w_1_"+tau_name).c_str(),
-            ("w_1_"+tau_name).c_str()
-        );
-        if (debug) hist_weights[tau_type]->SaveAs(("Temp_"+tau_name+".root").c_str()); // It's required that all bins are filled in these histograms; save them to check incase binning is too fine and some bins are empty
-
-        target_histogram.reset();
-        input_histogram .reset();
-      } 
+      for( auto const& [tau_type, tau_name] : tau_types_names){
+          if (tau_name != "data"){ // No histogram for data samples
```
Comment (kandrosov) [inline Training/interface/DataLoader_main.h:208]: here and in similar places below: don't use strings (string operation are much slower than integer comparison) -> use enum values

File: Training/interface/DataMixer.h [hunk 8]
```diff
@@ -1,15 +1,14 @@
...
     size_t n_per_batch;
-    std::string selection;
+    std::string selection; // selection for mixing
+    TH1D threshold; // threshold for event dropping
     std::vector<std::string> fileNames;
 };
 
-struct Dataset {
-    Dataset(const DatasetDesc& desc)  : desc_(desc) {
+struct DatasetMix {
```
Comment (kandrosov) [inline Training/interface/DataMixer.h:10]: use polymorphism
```c++
struct Dataset {
  virtual ~Dataset() {}
  virtual bool FillNext(input_tuple::TauTuple& outputTuple) = 0;
};

struct DatasetMix : Dataset {
  ...
  virtual bool FillNext(input_tuple::TauTuple& outputTuple) override
  {
    ...
  }
};

struct DatasetDrop : Dataset {
  ...
  virtual bool FillNext(input_tuple::TauTuple& outputTuple) override
  {
    ...
  }
};


class DataMixer {
public:
  DataMixer(const std::vector<DatasetDesc>& datasets_desc, const std::vector<OutputDesc>& outputDescs,
            const std::string& action)
    : outputDescs_(outputDescs)
  {
    if (action == "Mix"){
      CreateDatasets<DatasetMix>(datasets_desc);
    } else if(action == "Drop") {
      CreateDatasets<DatasetDrop>(datasets_desc);
    } else {
      throw std::runtime_error("Unsupported action");
    }
  }

  ...

private:
  template<typename DS>
  void CreateDatasets(const std::vector<DatasetDesc>& datasets_desc)
  {
    for(const auto& desc : datasets_desc) {
      std::shared_ptr<Dataset> ds = std::make_shared<DS>(desc);
      datasets_.push_back(ds);
    }
  }

private:
  std::vector<std::shared_ptr<Dataset>> datasets_;
};
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]
